File: java/com/google/copybara/util/Glob.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.util;

import static java.lang.Math.max;
import static java.lang.Math.min;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableCollection;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Iterables;
import java.nio.file.Path;
import java.nio.file.PathMatcher;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import javax.annotation.Nullable;
import net.starlark.java.annot.StarlarkBuiltin;
import net.starlark.java.eval.EvalException;
import net.starlark.java.eval.HasBinary;
import net.starlark.java.eval.Printer;
import net.starlark.java.eval.Starlark;
import net.starlark.java.eval.StarlarkValue;
import net.starlark.java.syntax.TokenKind;

/**
 * A {@link PathMatcher} builder that creates a PathMatcher relative to a {@link Path}.
 *
 * <p>The returned {@link PathMatcher} returns true if any of the {@code paths} expressions match.
 * If {@code paths} is empty it will no match any file.
 */
@StarlarkBuiltin(
    name = "glob",
    doc = "A glob represents a set of relative filepaths in the Copybara workdir.")
public class Glob implements StarlarkValue, HasBinary {

  public static final Glob ALL_FILES = createGlob(ImmutableList.of("**"));

  private final ImmutableList<GlobAtom> include;
  private final ImmutableList<Glob> globInclude;
  @Nullable private final Glob exclude;

  Glob(Iterable<GlobAtom> include, Iterable<Glob> globInclude, @Nullable Glob exclude) {
    this.include = ImmutableList.copyOf(Preconditions.checkNotNull(include));
    this.globInclude = ImmutableList.copyOf(Preconditions.checkNotNull(globInclude));
    this.exclude = exclude;
  }

  @Override
  public final Glob binaryOp(TokenKind op, Object that, boolean thisLeft) throws EvalException {
    switch (op) {
      case PLUS:
        if (that instanceof Glob) {
          return union(this, (Glob) that);
        } else {
          throw Starlark.errorf(
              "Cannot concatenate %s with %s. Only a glob can be concatenated to a glob",
              this, that);
        }
      case MINUS:
        if (that instanceof Glob) {
          return difference(this, (Glob) that);
        } else {
          throw Starlark.errorf(
              "Cannot subtract %s from %s. Only a glob can be subtracted from a glob", that, this);
        }
      default:
        throw Starlark.errorf("Glob does not support %s", op);
    }
  }

  /**
   * Compute the 'set union' of two Globs, which is a Glob that will match any Path matched by at
   * least one of those two Globs.
   *
   * <p>Try to keep the resulting Glob as flat as possible. In the worst case, this will increase
   * the depth of each leaf (i.e. {@code GlobAtom}) by 1, but we can often do better.
   */
  private static Glob union(Glob glob1, Glob glob2) {
    if (Objects.equals(glob1.exclude, glob2.exclude)) {
      return new Glob(
          Iterables.concat(glob1.include, glob2.include),
          Iterables.concat(glob1.globInclude, glob2.globInclude),
          glob1.exclude);
    } else if (glob1.exclude == null) {
      return new Glob(
          glob1.include, Iterables.concat(glob1.globInclude, ImmutableList.of(glob2)), null);
    } else if (glob2.exclude == null) {
      return new Glob(
          glob2.include, Iterables.concat(glob2.globInclude, ImmutableList.of(glob1)), null);
    }
    return new Glob(ImmutableList.of(), ImmutableList.of(glob1, glob2), null);
  }

  /**
   * Compute the 'set difference' of two Globs, which is a Glob that will match any Path which is
   * matched by the first Glob, but not matched by the second Glob.
   */
  private static Glob difference(Glob glob1, Glob glob2) {
    if (glob1.exclude == null) {
      return new Glob(glob1.include, glob1.globInclude, glob2);
    }
    return new Glob(glob1.include, glob1.globInclude, Glob.union(glob1.exclude, glob2));
  }

  /** Checks if the given {@code changedFiles} are or are descendants of the {@code roots}. */
  public static boolean affectsRoots(
      ImmutableSet<String> roots, ImmutableCollection<String> changedFiles) {
    if (changedFiles == null || isEmptyRoot(roots)) {
      return true;
    }
    // This is O(changes * files * roots) in the worse case. roots shouldn't be big and
    // files shouldn't be big for 99% of the changes.
    for (String file : changedFiles) {
      for (String root : roots) {
        if (file.equals(root) || file.startsWith(root + "/")) {
          return true;
        }
      }
    }
    return false;
  }

  public PathMatcher relativeTo(Path path) {
    ImmutableList.Builder<PathMatcher> includeList = ImmutableList.builder();
    for (GlobAtom path1 : include) {
      includeList.add(path1.matcher(path));
    }
    for (Glob g : globInclude) {
      includeList.add(g.relativeTo(path));
    }
    PathMatcher excludeMatcher =
        (exclude == null) ? FileUtil.anyPathMatcher(ImmutableList.of()) : exclude.relativeTo(path);
    return new GlobPathMatcher(FileUtil.anyPathMatcher(includeList.build()), excludeMatcher);
  }

  private class GlobPathMatcher implements PathMatcher {

    private final PathMatcher includeMatcher;
    private final PathMatcher excludeMatcher;

    GlobPathMatcher(PathMatcher includeMatcher, PathMatcher excludeMatcher) {
      this.includeMatcher = includeMatcher;
      this.excludeMatcher = excludeMatcher;
    }

    @Override
    public boolean matches(Path path) {
      return includeMatcher.matches(path) && !excludeMatcher.matches(path);
    }

    @Override
    public String toString() {
      return Glob.this.toString();
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (!(o instanceof GlobPathMatcher)) {
        return false;
      }
      GlobPathMatcher that = (GlobPathMatcher) o;
      return Objects.equals(this.includeMatcher, that.includeMatcher)
          && Objects.equals(this.excludeMatcher, that.excludeMatcher);
    }

    @Override
    public int hashCode() {
      return Objects.hash(includeMatcher, excludeMatcher);
    }
  }

  /**
   * Creates a function {@link Glob} that when a {@link Path} is passed it returns a {@link
   * PathMatcher} relative to the path.
   *
   * @param include list of strings representing the globs to include/match
   * @param exclude list of strings representing the globs to exclude from the include set
   * @throws IllegalArgumentException if any glob is not valid
   */
  public static Glob createGlob(Iterable<String> include, Iterable<String> exclude) {
    return new Glob(
        ImmutableList.copyOf(GlobAtom.ofIterable(include, GlobAtom.AtomType.JAVA_GLOB)),
        ImmutableList.of(),
        Iterables.isEmpty(exclude) ? null : createGlob(exclude));
  }

  /**
   * Creates a function {@link Glob} that when a {@link Path} is passed it returns a {@link
   * PathMatcher} relative to the path.
   *
   * @param include list of strings representing the globs to include/match
   * @throws IllegalArgumentException if any glob is not valid
   */
  public static Glob createGlob(Iterable<String> include) {
    return createGlob(include, ImmutableList.of());
  }

  /**
   * Calculates a set of paths which recursively contain all files that could possibly match a file
   * in this glob. Generally, this returns the include paths but removing all segments that have a
   * metacharacter and following segments.
   *
   * <p>This can be used by an {@code Origin} or {@code Destination} implementation to determine
   * which directories to query from the repo. For instance, if the <em>include</em> paths are:
   *
   * <ul>
   *   <li>foo/bar.jar
   *   <li>foo/baz/**
   * </ul>
   *
   * This function will return a single string: {@code "foo"}.
   *
   * <p>If the include paths potentially include files in the root directory or use metacharacters
   * to specify the top level directory, a set with only the empty string is returned. For instance,
   * the following include globs will cause {@code [""]} (and no other string) to be returned:
   *
   * <ul>
   *   <li>{@code *.java}
   *   <li>{@code {foo,bar}/baz/**}
   * </ul>
   *
   * <p>Note that in the case of {@code origin_files} or {@code destination_files}, the origin or
   * destination may give special meaning to the roots of the glob. For instance, the destination
   * may store metadata at {root}/INFO for every root. See the documentation of the destination or
   * origin you are using for more information.
   */
  public ImmutableSet<String> roots() {
    return roots(false);
  }

  /**
   * If {@code allowFiles} is set to true, then Paths containing no meta characters are retained
   * exactly as they are - for example, {@code foo/bar.txt} is output unmodified and not shortened
   * to {@code foo}.
   */
  public ImmutableSet<String> roots(boolean allowFiles) {
    return computeRootsFromIncludes(getIncludes(), allowFiles);
  }

  /**
   * If roots is empty or contains a single elemnent that is not a subdirectory. See {@link
   * #roots()} for detail.
   */
  public static boolean isEmptyRoot(Iterable<String> roots) {
    return Iterables.isEmpty(roots) || Objects.equals(roots.iterator().next(), "");
  }

  protected Iterable<GlobAtom> getIncludes() {
    return Iterables.concat(
        include, Iterables.concat(Iterables.transform(globInclude, Glob::getIncludes)));
  }

  private static ImmutableSet<String> computeRootsFromIncludes(
      Iterable<GlobAtom> includes, boolean allowFiles) {
    List<String> roots = new ArrayList<>();

    for (GlobAtom atom : includes) {
      roots.add(atom.root(allowFiles));
    }

    // Remove redundant roots - e.g. "foo" covers all paths that start with "foo/"
    Collections.sort(roots, Glob::compareRoots);
    if (roots.contains("")) {
      return ImmutableSet.of("");
    }
    int r = 0;
    while (r < roots.size() - 1) {
      if (roots.get(r + 1).startsWith(roots.get(r) + "/")) {
        roots.remove(r + 1);
      } else {
        r++;
      }
    }

    return ImmutableSet.copyOf(roots);
  }

  /** A lexicographical String comparator that sorts the '/' char before any other char. */
  private static int compareRoots(String s1, String s2) {
    int len1 = s1.length();
    int len2 = s2.length();
    int lim = min(len1, len2);
    for (int k = 0; k < lim; k++) {
      int c1 = s1.charAt(k);
      c1 = c1 == '/' ? -1 : c1;
      int c2 = s2.charAt(k);
      c2 = c2 == '/' ? -1 : c2;
      if (c1 != c2) {
        return c1 - c2;
      }
    }
    return len1 - len2;
  }

  @Override
  public String toString() {
    return toStringWithParentheses(true);
  }

  private String toStringWithParentheses(boolean isRootGlob) {
    StringBuilder builder = new StringBuilder();
    int numberOfTerms = 0;
    boolean inlineExclude =
        this.globInclude.isEmpty()
            && this.exclude != null
            && this.exclude.globInclude.isEmpty()
            && this.exclude.exclude == null;
    if (!include.isEmpty() || this.globInclude.isEmpty()) {
      builder
          .append("glob(include = ")
          .append(toStringList(include))
          .append(inlineExclude ? ", exclude = " + toStringList(this.exclude.include) : "")
          .append(")");
      numberOfTerms += 1;
    }
    for (Glob g : this.globInclude) {
      if (numberOfTerms > 0) {
        builder.append(" + ");
      }
      builder.append(g.toStringWithParentheses(false));
      numberOfTerms += 1;
    }
    if (this.exclude != null && !inlineExclude) {
      builder.append(" - ").append(this.exclude.toStringWithParentheses(false));
      numberOfTerms += 1;
    }

    if (!isRootGlob && numberOfTerms > 1) {
      return "(" + builder + ")";
    }
    return builder.toString();
  }

  private String toStringList(Iterable<GlobAtom> iterable) {
    StringBuilder sb = new StringBuilder("[");
    boolean first = true;
    for (GlobAtom atom : iterable) {
      if (first) {
        first = false;
      } else {
        sb.append(", ");
      }
      sb.append('"').append(sanitize(atom.toString())).append('"');
    }
    return sb.append("]").toString();
  }

  private String sanitize(String s) {
    return s.replace("\\", "\\\\")
        .replace("\"", "\\\"")
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
        .replace("\f", "\\f")
        .replace("\b", "\\b")
        .replace("\000", "\\000");
  }

  @Override
  public void repr(Printer printer) {
    printer.append(toString());
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof Glob)) {
      return false;
    }
    Glob that = (Glob) o;
    return Objects.equals(include, that.include)
        && Objects.equals(globInclude, that.globInclude)
        && Objects.equals(exclude, that.exclude);
  }

  @Override
  public int hashCode() {
    return Objects.hash(include, globInclude, exclude);
  }

  @VisibleForTesting
  public int heightOfGlobTree() {
    int includeHeight = this.globInclude.stream().mapToInt(Glob::heightOfGlobTree).max().orElse(-1);
    int excludeHeight = this.exclude == null ? -1 : this.exclude.heightOfGlobTree();
    return 1 + max(includeHeight, excludeHeight);
  }
}

File: src/Copybara.Core/Util/Glob.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using System.Text;
using Copybara.Common;
using Starlark.Annot;
using Starlark.Eval;
using StarlarkRuntime = Starlark.Eval.Starlark;

namespace Copybara.Util;

/// <summary>
/// A <see cref="IPathMatcher"/> builder that creates a matcher relative to a root path. Port of
/// <c>com.google.copybara.util.Glob</c>.
///
/// <para>The returned <see cref="IPathMatcher"/> returns true if any of the <c>paths</c> expressions
/// match. If <c>paths</c> is empty it will not match any file.</para>
/// </summary>
[StarlarkBuiltin(
    "glob",
    Doc = "A glob represents a set of relative filepaths in the Copybara workdir. Most consumers "
        + "will also accept a list of fully qualified (no wildcards) file names instead.")]
public class Glob : IStarlarkValue, IEquatable<Glob>
{
    public static readonly Glob AllFiles = CreateGlob(ImmutableArray.Create("**"));

    protected readonly ImmutableArray<GlobAtom> Include;
    private readonly ImmutableArray<Glob> _globInclude;
    private readonly Glob? _exclude;

    internal Glob(IEnumerable<GlobAtom> include, IEnumerable<Glob> globInclude, Glob? exclude)
    {
        Include = Preconditions.CheckNotNull(include).ToImmutableArray();
        _globInclude = Preconditions.CheckNotNull(globInclude).ToImmutableArray();
        _exclude = exclude;
    }

    /// <summary>
    /// Implements the Starlark <c>+</c> (union / list concatenation) and <c>-</c> (difference)
    /// operators. Port of Java's <c>binaryOp</c>.
    /// </summary>
    public Glob BinaryOp(string op, object that, bool thisLeft)
    {
        switch (op)
        {
            case "+":
                if (that is Glob addGlob)
                {
                    return Union(this, addGlob);
                }
                if (that is IEnumerable<string> addList)
                {
                    return new Glob(
                        ImmutableArray<GlobAtom>.Empty,
                        new[] { this, SequenceGlob.OfIterable(addList) },
                        null);
                }
                throw StarlarkRuntime.Errorf(
                    "Cannot concatenate {0} with {1}. Only a glob can be concatenated to a glob",
                    this, that);
            case "-":
                if (that is Glob subGlob)
                {
                    return Difference(this, subGlob);
                }
                if (that is IEnumerable<string> subList)
                {
                    Glob list = SequenceGlob.OfIterable(subList);
                    return new Glob(
                        Include, _globInclude, _exclude != null ? Union(_exclude, list) : list);
                }
                throw StarlarkRuntime.Errorf(
                    "Cannot subtract {0} from {1}. Only a glob can be subtracted from a glob", that, this);
            default:
                throw StarlarkRuntime.Errorf("Glob does not support {0}", op);
        }
    }

    /// <summary>
    /// Compute the 'set union' of two Globs, which is a Glob that will match any Path matched by at
    /// least one of those two Globs.
    /// </summary>
    public static Glob Union(Glob glob1, Glob glob2)
    {
        if (Equals(glob1._exclude, glob2._exclude))
        {
            return new Glob(
                glob1.Include.Concat(glob2.Include),
                glob1._globInclude.Concat(glob2._globInclude),
                glob1._exclude);
        }
        if (glob1._exclude == null)
        {
            return new Glob(
                glob1.Include, glob1._globInclude.Concat(new[] { glob2 }), null);
        }
        if (glob2._exclude == null)
        {
            return new Glob(
                glob2.Include, glob2._globInclude.Concat(new[] { glob1 }), null);
        }
        return new Glob(ImmutableArray<GlobAtom>.Empty, new[] { glob1, glob2 }, null);
    }

    /// <summary>
    /// Compute the 'set difference' of two Globs, which is a Glob that will match any Path which is
    /// matched by the first Glob, but not matched by the second Glob.
    /// </summary>
    public static Glob Difference(Glob glob1, Glob glob2)
    {
        if (glob1._exclude == null)
        {
            return new Glob(glob1.Include, glob1._globInclude, glob2);
        }
        return new Glob(glob1.Include, glob1._globInclude, Union(glob1._exclude, glob2));
    }

    /// <summary>Checks if the given <paramref name="changedFiles"/> are or are descendants of the roots.</summary>
    public static bool AffectsRoots(
        ImmutableHashSet<string> roots, IReadOnlyCollection<string>? changedFiles)
    {
        if (changedFiles == null || IsEmptyRoot(roots))
        {
            return true;
        }
        foreach (var file in changedFiles)
        {
            foreach (var root in roots)
            {
                if (file == root || file.StartsWith(root + "/"))
                {
                    return true;
                }
            }
        }
        return false;
    }

    public virtual IPathMatcher RelativeTo(string path)
    {
        var includeList = new List<IPathMatcher>();
        foreach (var atom in Include)
        {
            includeList.Add(atom.Matcher(path));
        }
        foreach (var g in _globInclude)
        {
            includeList.Add(g.RelativeTo(path));
        }
        IPathMatcher excludeMatcher =
            _exclude == null ? FileUtil.AnyPathMatcher(ImmutableArray<IPathMatcher>.Empty) : _exclude.RelativeTo(path);
        return new GlobMatcher(FileUtil.AnyPathMatcher(includeList), excludeMatcher, ToString());
    }

    private sealed class GlobMatcher : IPathMatcher, IEquatable<GlobMatcher>
    {
        private readonly IPathMatcher _includeMatcher;
        private readonly IPathMatcher _excludeMatcher;
        private readonly string _toString;

        public GlobMatcher(IPathMatcher includeMatcher, IPathMatcher excludeMatcher, string toString)
        {
            _includeMatcher = includeMatcher;
            _excludeMatcher = excludeMatcher;
            _toString = toString;
        }

        public bool Matches(string path) =>
            _includeMatcher.Matches(path) && !_excludeMatcher.Matches(path);

        public override string ToString() => _toString;

        public bool Equals(GlobMatcher? other) =>
            other is not null
            && Equals(_includeMatcher, other._includeMatcher)
            && Equals(_excludeMatcher, other._excludeMatcher);

        public override bool Equals(object? obj) => Equals(obj as GlobMatcher);

        public override int GetHashCode() => HashCode.Combine(_includeMatcher, _excludeMatcher);
    }

    /// <summary>Creates a <see cref="Glob"/> from include and exclude patterns.</summary>
    public static Glob CreateGlob(IEnumerable<string> include, IEnumerable<string> exclude)
    {
        var excludeList = exclude as IReadOnlyCollection<string> ?? exclude.ToList();
        return new Glob(
            GlobAtom.OfIterable(include, GlobAtom.AtomType.JavaGlob).ToImmutableArray(),
            ImmutableArray<Glob>.Empty,
            excludeList.Count == 0 ? null : CreateGlob(excludeList));
    }

    /// <summary>Creates a <see cref="Glob"/> that matches the given single file paths.</summary>
    public static Glob CreateSingleFilesGlob(IEnumerable<string> singleFilePaths) =>
        SequenceGlob.OfIterable(singleFilePaths);

    /// <summary>Handles a Glob/Sequence parameter passed to a Glob-consuming function.</summary>
    public static Glob? WrapGlob(object? globOrList, Glob? defaultValue)
    {
        if (globOrList is null || ReferenceEquals(globOrList, StarlarkRuntime.None))
        {
            return defaultValue;
        }
        if (globOrList is Glob glob)
        {
            return glob;
        }
        if (globOrList is IEnumerable<string> list)
        {
            return SequenceGlob.OfIterable(list);
        }
        throw new EvalException("Glob can only be created from a Glob or a list of strings");
    }

    /// <summary>Creates a <see cref="Glob"/> from include patterns only.</summary>
    public static Glob CreateGlob(IEnumerable<string> include) =>
        CreateGlob(include, ImmutableArray<string>.Empty);

    /// <summary>
    /// Calculates a set of paths which recursively contain all files that could possibly match a file
    /// in this glob. See the Java documentation for the exact semantics.
    /// </summary>
    public ImmutableHashSet<string> Roots() => Roots(false);

    /// <summary>Similar to <see cref="Roots()"/> but returns the longest shared paths for git cones.</summary>
    public ImmutableHashSet<string> Tips() => ComputeTipsFromIncludes(GetIncludes());

    /// <summary>
    /// If <paramref name="allowFiles"/> is set, then paths containing no meta characters are retained
    /// exactly as they are.
    /// </summary>
    public ImmutableHashSet<string> Roots(bool allowFiles) =>
        ComputeRootsFromIncludes(GetIncludes(), allowFiles);

    /// <summary>If roots is empty or contains a single element that is not a subdirectory.</summary>
    public static bool IsEmptyRoot(IEnumerable<string> roots)
    {
        using var it = roots.GetEnumerator();
        return !it.MoveNext() || it.Current == "";
    }

    protected virtual IEnumerable<GlobAtom> GetIncludes() =>
        Include.Concat(_globInclude.SelectMany(g => g.GetIncludes()));

    private static ImmutableHashSet<string> ComputeRootsFromIncludes(
        IEnumerable<GlobAtom> includes, bool allowFiles)
    {
        var roots = new List<string>();
        foreach (var atom in includes)
        {
            roots.Add(atom.Root(allowFiles));
        }

        roots.Sort(CompareRoots);
        if (roots.Contains(""))
        {
            return ImmutableHashSet.Create("");
        }
        int r = 0;
        while (r < roots.Count - 1)
        {
            if (roots[r + 1].StartsWith(roots[r] + "/"))
            {
                roots.RemoveAt(r + 1);
            }
            else
            {
                r++;
            }
        }

        return roots
            .Select(s => CollapseSlashes(s))
            .ToImmutableHashSet();
    }

    private static string CollapseSlashes(string s)
    {
        // Replace "//+" with "/" then remove a trailing "/".
        var sb = new StringBuilder(s.Length);
        bool prevSlash = false;
        foreach (var c in s)
        {
            if (c == '/')
            {
                if (!prevSlash)
                {
                    sb.Append('/');
                }
                prevSlash = true;
            }
            else
            {
                sb.Append(c);
                prevSlash = false;
            }
        }
        string result = sb.ToString();
        if (result.EndsWith('/'))
        {
            result = result.Substring(0, result.Length - 1);
        }
        return result;
    }

    private static ImmutableHashSet<string> ComputeTipsFromIncludes(IEnumerable<GlobAtom> includes)
    {
        var wildcards = new List<string>();
        var singleFiles = new List<string>();
        foreach (var atom in includes)
        {
            Root root = atom.AnnotatedRoot(false);
            if (!root.IsRecursive())
            {
                singleFiles.Add(root.GetRoot());
            }
            else
            {
                wildcards.Add(root.GetRoot());
            }
        }
        wildcards.Sort(CompareRoots);
        if (wildcards.Contains(""))
        {
            return ImmutableHashSet.Create("");
        }
        int r = 0;
        while (r < wildcards.Count - 1)
        {
            if (wildcards[r + 1].StartsWith(wildcards[r] + "/"))
            {
                wildcards.RemoveAt(r + 1);
            }
            else
            {
                r++;
            }
        }

        singleFiles.Sort((a, b) => CompareRoots(b, a));
        int t = 0;
        while (t < singleFiles.Count - 1)
        {
            if (singleFiles[t].StartsWith(singleFiles[t + 1] + "/"))
            {
                singleFiles.RemoveAt(t + 1);
            }
            else
            {
                t++;
            }
        }
        var tips = new List<string>(wildcards);

        foreach (var single in singleFiles)
        {
            if (!wildcards.Any(w => (single + "/").StartsWith(w + "/")))
            {
                tips.Add(single);
            }
        }
        tips.AddRange(wildcards);
        return tips.ToImmutableHashSet();
    }

    /// <summary>A lexicographical string comparator that sorts the '/' char before any other char.</summary>
    private static int CompareRoots(string s1, string s2)
    {
        int len1 = s1.Length;
        int len2 = s2.Length;
        int lim = Math.Min(len1, len2);
        for (int k = 0; k < lim; k++)
        {
            int c1 = s1[k];
            c1 = c1 == '/' ? -1 : c1;
            int c2 = s2[k];
            c2 = c2 == '/' ? -1 : c2;
            if (c1 != c2)
            {
                return c1 - c2;
            }
        }
        return len1 - len2;
    }

    public override string ToString() => ToStringWithParentheses(true);

    internal virtual string ToStringWithParentheses(bool isRootGlob)
    {
        var builder = new StringBuilder();
        int numberOfTerms = 0;
        bool inlineExclude =
            _globInclude.IsEmpty
            && _exclude != null
            && _exclude.GetType() == GetType()
            && _exclude._globInclude.IsEmpty
            && _exclude._exclude == null;
        if (!Include.IsEmpty || _globInclude.IsEmpty)
        {
            builder
                .Append("glob(include = ")
                .Append(ToStringList(Include))
                .Append(inlineExclude ? ", exclude = " + ToStringList(_exclude!.Include) : "")
                .Append(')');
            numberOfTerms += 1;
        }
        foreach (var g in _globInclude)
        {
            if (numberOfTerms > 0)
            {
                builder.Append(" + ");
            }
            builder.Append(g.ToStringWithParentheses(false));
            numberOfTerms += 1;
        }
        if (_exclude != null && !inlineExclude)
        {
            builder.Append(" - ").Append(_exclude.ToStringWithParentheses(false));
            numberOfTerms += 1;
        }

        if (!isRootGlob && numberOfTerms > 1)
        {
            return "(" + builder + ")";
        }
        return builder.ToString();
    }

    internal string ToStringList(IEnumerable<GlobAtom> iterable)
    {
        var sb = new StringBuilder("[");
        bool first = true;
        foreach (var atom in iterable)
        {
            if (first)
            {
                first = false;
            }
            else
            {
                sb.Append(", ");
            }
            sb.Append('"').Append(Sanitize(atom.ToString())).Append('"');
        }
        return sb.Append(']').ToString();
    }

    private static string Sanitize(string s) =>
        s.Replace("\\", "\\\\")
            .Replace("\"", "\\\"")
            .Replace("\n", "\\n")
            .Replace("\r", "\\r")
            .Replace("\t", "\\t")
            .Replace("\f", "\\f")
            .Replace("\b", "\\b")
            .Replace("\0", "\\000");

    public bool Equals(Glob? other) =>
        other is not null
        && Include.SequenceEqual(other.Include)
        && _globInclude.SequenceEqual(other._globInclude)
        && Equals(_exclude, other._exclude);

    public override bool Equals(object? obj) => Equals(obj as Glob);

    public override int GetHashCode()
    {
        var hc = new HashCode();
        foreach (var a in Include)
        {
            hc.Add(a);
        }
        foreach (var g in _globInclude)
        {
            hc.Add(g);
        }
        hc.Add(_exclude);
        return hc.ToHashCode();
    }

    public int HeightOfGlobTree()
    {
        int includeHeight = _globInclude.Length == 0
            ? -1
            : _globInclude.Max(g => g.HeightOfGlobTree());
        int excludeHeight = _exclude == null ? -1 : _exclude.HeightOfGlobTree();
        return 1 + Math.Max(includeHeight, excludeHeight);
    }
}
